File: Python/functs_Idolor.py
import re 

############################ new_user function ###############################
def new_user(local_hash): 

	# prompt for username 
	print ("Please Enter a Username\n")
	username  = input()
	# strip off special chars from username (ref asmt sheet)
	username = re.sub(r'\W+','',username)
	# convert username to lowercase (ref asmt sheet)
	username = username.lower()
	#check if username already exists & if so exit, else assign password as value to the hash referenced by username as the key
	if (username in local_hash):
	 print ("Username name already exisits") 
	 return local_hash
	print ("Please Enter a Password")
	password = input()
	# strip off apostrophe from password (ref asmt sheet)
	password = password.replace('\'', '')
	local_hash[username] = password
	return local_hash
	
	
	
	

############################ edit_user function ###############################
def edit_user(local_hash):

	# prompt for username 
	print ("Please Enter a Username\n")
	username  = input()
	# strip off special chars from username (ref asmt sheet)
	username = re.sub(r'\W+','',username)
	# convert username to lowercase (ref asmt sheet)
	username = username.lower()
	#check if username already exists & if so exit, else assign password as value to the hash referenced by username as the key
	if (username not in local_hash):
	 print ("Username name does not exisit") 
	 return local_hash
	#prompt/chomp for current password
	print ("Enter Password to edit account")
	password = input() # strip off apostrophe from password (ref asmt sheet)
	password = password.replace('\'', '')
	#check if password DOESNT exist & if NOT so exit
	if (local_hash[username] == password) : 
	#prompt/chomp for new password
	 print ("Enter New Password" ) 
	 newpassword = input() 
	# strip off apostrophe from password (ref asmt sheet)
	 newpassword = newpassword.replace('\'', '')
	# assign password as value to the hash referenced by username as the key
	 local_hash[username] = newpassword
	else : 
	 return local_hash
	return local_hash 

############################ delete_user function ###############################
def delete_user(local_hash):

	# prompt/chomp for username
	print ("Enter Username to delete") 
	username = input() 
	# strip off special chars from username (ref asmt sheet)
	username = re.sub(r'\W+','',username)
	# convert username to lowercase (ref asmt sheet)
	username = username.lower()
	#check if username DOESNT already exists & if NOT so exit
	if (username in local_hash):
	#delete key from hash (ref slides)
	 del (local_hash[username] )
	 print ("User deleted") 
	 return local_hash
	else: 
	 print ("Username does not exist") 
	return local_hash

############################ list_accounts function ###############################
def list_accounts(local_hash):

	# print out to the screen each key & value pair from hash
	for k,v in local_hash.items():
	 print(k + ":" + v + "\n" ) 
	return local_hash

File: Perl/functs_Idolor.pl
 use strict; 
 use functs_Idolor; 

# create a hash of functions
my %function = ('n' =>\ &new_user, 'e' =>\ &edit_user , 'd' => \ &delete_user, 'l' => \ &list_accounts) ; 
#global var declarations go here
my %hash; 
my $choice;
#Open file for reading
# prompt user for filename & store in variable:"fn"  : print, <>, chomp
print "What File Would you Like to Open\n"; 
my $fn = <>; 
chomp ($fn); 
open (my $fh, "<",  $fn ); 
# if file does not exist, create it
if (!$fh) {

open (my $fh , ">", $fn ) ; 
	} 
	
#transfer file line by line into hash with a while loop, splitting the username and password by the delimiter character ":"
while (my $line = <$fh>) {

my @line_array = split(/:/ , $line ) ; 
my $username =  $line_array[0] ;
my $password =  $line_array[1] ; 
chomp ($password ); 
 $hash{$username} = $password; 
} 
# outer loop until user quits
outer_loop: while (1) {
	#inner loop until user enters correct option
inner_loop:  while (1) {

	# print menu & prompt/chomp for user input: $choice
	
	print "User Accounts\n";
	print "----------------\n";
	print " n = New User Account\n";
	print " e = Edit Existing User Account\n";
	print " d = Delete Exisiting User Account\n" ;
	print " l = List User Accounts\n";
	print " q = Quit \n";
	
	print "Enter Choice \n ";
	$choice = <>; 
	chomp ($choice);
# is uer enters 'q' then quit by exiting outer loop
if ($choice eq 'q' ) {
last outer_loop;
}
# if the user enters a valid choice, exit inner loop	
if (exists ($function {$choice} )) { last inner_loop; } 

print "Unknown Choice\n"; 

# set local variable to appropriate function references in hash by user's choice & call appropriate function by passing in the hash as a parameter and assigning the function result to the hash to update it
}
my $selection =  $function {$choice} ; 
%hash = $selection -> (\%hash); 
}

#Close the file for reading
close ($fh); 
# prompt user to save changes, if so, then print the entire contents of the hash to the file & close the file for writing

print "Save Changes?(y/n)?\n";
my $saveit = <>; 
chomp($saveit); 
if ( $saveit eq ('y')){ 
open ($fh,">", $fn); 
for my $key(keys %hash ) {
print $fh $key.":". $hash {$key}."\n"; 
}
close ($fh); 
print "Usernames and Passwords saved\n";
}